documentation/typescript/adding-custom-login-form-with-spring-security.asciidoc

```diff
@@ -0,0 +1,176 @@
+    // Vaadin already handles csrf.
+    http.csrf().disable();
+    // specify the URL of the login view, the value of the parameter
+    // is the defined route for the login view component.
+    http.formLogin().loginPage("login");
+  }  
+}
+----
+
+== Custom login view
+
+The easiest way to make a login view is to use the `<vaadin-login-overlay>` component.
+Vaadin provides a `login` helper method for Spring Security based authentication that you can use for the login action.
+
+.frontend/login-view.ts
+[source, typescript]
+----
+@customElement('login-view')
+export class LoginView extends LitElement implements AfterEnterObserver {
+
+  @property({type: Boolean})
+  private error = false;
+
+  @property({type: Boolean})
+  private open = true;
+
+  // the url to redirect to after a successful login
+  private returnUrl = '/';
+
+  private onSuccess: (result:LoginResult) => void;
+
+  constructor(){
+    super();
+    this.onSuccess = () => {
+      Router.go(this.returnUrl);
+    };
+  }
+
+  render() {
+    return html`
+      <vaadin-login-overlay
+        ?opened="${this.open}" 
+        .error=${this.error}
+        @login="${this.login}">    
+      </vaadin-login-overlay>
+    `;
+  }
+
+  async login(event: CustomEvent): Promise<LoginResult> {
+    this.error = false;
+    // use the Vaadin provided login helper method to obtain the LoginResult
+    const result = await login(event.detail.username, event.detail.password);
+    this.error = result.error;
+  
+    if (!result.error) {
+      this.onSuccess(result);
+    }
+
+    return result;
+  }
+
+  onAfterEnter(location: RouterLocation) {
+    this.returnUrl = location.redirectFrom || this.returnUrl;
+  }
+}
+----
+After the login view is defined, you should also define a route for the login view component, for example in the `index.ts` file. Note, the `path` for the login view component should be the same as what you defined in `SecurityConfig` for `http.formLogin().loginPage()`.
+.frontend/index.ts
+[source, typescript]
+----
+const routes = [
+  {
+    path: '/login',
+    component: 'login-view'
+  },
+  // more routes
+}
+----
+== Redirect the user to the login view
+To protect a view from unauthenticated users, that is, redirect unauthenticated users to the login view, you can use route action.
+
+.frontend/index.ts
+[source, typescript]
+----
+const isUserLoggedIn = function() {
+  return localStorage.getItem('loggedIn') === 'true';
+}
+
+const routes = [
+  ...
+  {
+    path: '/my-view',
+    action: async (_: Router.Context, commands: Router.Commands) => {
+      if (!await isUserLoggedIn()) {
+        return commands.redirect('/login');
+      }
+      return undefined;
+    },
+    component: 'my-view'
+  }
+  ...
+}
+----
+You can also add the route action to the parent layout, so that all the child views are protected.
+
+.frontend/index.ts
+[source, typescript]
+----
+const routes = [
+  ...
+  {
+    path: '/main-layout',
```

Comment:
The main layout is supposed to apply to the entire application, maybe this `path` should be `'/'` accordingly.

Reply:
good catch, fixed